crates/autopilot/src/run_loop.rs

```diff
@@ -635,21 +635,37 @@ impl RunLoop {
         request: &solve::Request,
     ) -> Result<Vec<Result<competition::Solution, domain::competition::SolutionError>>, SolveError>
     {
-        let can_participate = self.solver_participation_guard.can_participate(&driver.submission_address).await.map_err(|err| {
-            tracing::error!(?err, driver = %driver.name, ?driver.submission_address, "solver participation check failed");
-                SolveError::SolverDenyListed
+        let timeout = tokio::time::sleep(self.config.solve_deadline).fuse();
+        let send_request = driver.solve(request).fuse();
+        let check_allowed = self
+            .solver_participation_guard
+            .can_participate(&driver.submission_address)
+            .fuse();
+        tokio::pin!(timeout, send_request, check_allowed);
+
+        let response = loop {
+            tokio::select! {
+                _ = &mut timeout => return Err(SolveError::Timeout),
+                is_allowed = &mut check_allowed => match is_allowed {
+                    Ok(false) => return Err(SolveError::SolverDenyListed),
+                    Ok(true) => (), // all good => continue with other tasks
```

Comment:
Wouldn't that require something like this?
```suggestion
	            Ok(true) => {
	                // Continue waiting for response or timeout
	                tokio::select! {
	                    _ = &mut timeout => return Err(SolveError::Timeout),
	                    response = &mut send_request => match response {
	                        Ok(r) => r,
	                        Err(e) => return Err(SolveError::Failure(e)),
	                    },
	                }
	            },
```

Reply:
If we do this we'd have to duplicate the logic for the timeout and the request. Note that we `loop` over the `select!()` macro. That means if the `Ok(true)` case does nothing we redo the select and now we can only make progress on the other 2 macros.
I replaced the `()` with `continue` to make it more explicit what's going on.